File: middle_node_of_linked_list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var middleNode = function(head) {
    let oneStepper = head;
    let twoStepper = head;

    while (twoStepper){
        if (twoStepper.next == null){
            break;
        }
        oneStepper = oneStepper.next;
        twoStepper = twoStepper.next.next;
    }
    return oneStepper;
};

File: middle_node_of_linked_list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def middleNode(self, head: Optional[ListNode]) -> Optional[ListNode]:
        curr_oneStep = head
        curr_twoStep = head

        while curr_twoStep:

            if curr_twoStep.next == None:
                break
            curr_twoStep = curr_twoStep.next.next
            curr_oneStep = curr_oneStep.next
        
        return curr_oneStep